src/jit/codegenarm64.cpp

```diff
@@ -2354,6 +2354,315 @@ void CodeGen::genCodeForBinary(GenTree* treeNode)
+        assert(lclType == TYP_STRUCT);  
+        assert(varDsc->lvIsMultiRegRet);
+
+        ReturnTypeDesc  retTypeDesc;
+        unsigned        regCount;
+
+        retTypeDesc.InitializeStructReturnType(compiler, varDsc->lvVerTypeInfo.GetClassHandle());
+        regCount = retTypeDesc.GetReturnRegCount();
+
+        assert(regCount >= 2);
+        assert(op1->isContained());
+
+        // Copy var on stack into ABI return registers     
+        int offset = 0;
+        for (unsigned i = 0; i < regCount; ++i)
+        {
+            var_types type = retTypeDesc.GetReturnRegType(i);
+            regNumber reg = retTypeDesc.GetABIReturnReg(i);
+            getEmitter()->emitIns_R_S(ins_Load(type), emitTypeSize(type), reg, lclVar->gtLclNum, offset);
+            offset += genTypeSize(type);
+        }
+    }
+    else // op1 must be multi-reg GT_CALL
+    {
+        assert(op1->IsMultiRegCall() || op1->IsCopyOrReloadOfMultiRegCall());
+
+        genConsumeRegs(op1);
+
+        GenTree* actualOp1 = op1->gtSkipReloadOrCopy();
+        GenTreeCall* call = actualOp1->AsCall();
+
+        ReturnTypeDesc*  pRetTypeDesc;
+        unsigned         regCount;
+        unsigned         matchingCount = 0;
+
+        pRetTypeDesc = call->GetReturnTypeDesc();
+        regCount     = pRetTypeDesc->GetReturnRegCount();
+
+        var_types  regType     [MAX_RET_REG_COUNT];
+        regNumber  returnReg   [MAX_RET_REG_COUNT];
+        regNumber  allocatedReg[MAX_RET_REG_COUNT];
+        regMaskTP  srcRegsMask = 0;
+        regMaskTP  dstRegsMask = 0;
+        bool       needToShuffleRegs = false;   // Set to true if we have to move any registers
+
+        for (unsigned i = 0; i < regCount; ++i)
+        {
+            regType[i]      = pRetTypeDesc->GetReturnRegType(i);
+            returnReg[i]    = pRetTypeDesc->GetABIReturnReg(i);
+
+            regNumber reloadReg = REG_NA;
+            if (op1->IsCopyOrReload())
+            {
+                // GT_COPY/GT_RELOAD will have valid reg for those positions
+                // that need to be copied or reloaded.
+                reloadReg = op1->AsCopyOrReload()->GetRegNumByIdx(i);
+            }
+
+            if (reloadReg != REG_NA)
+            {
+                allocatedReg[i] = reloadReg;
+            }
+            else
+            {
+                allocatedReg[i] = call->GetRegNumByIdx(i);
+            }
+
+            if (returnReg[i] == allocatedReg[i])
+            {
+                matchingCount++;
+            }
+            else // We need to move this value
+            {
+                // We want to move the value from allocatedReg[i] into returnReg[i]
+                // so record these two registers in the src and dst masks
+                //
+                srcRegsMask |= genRegMask(allocatedReg[i]);
+                dstRegsMask |= genRegMask(returnReg[i]);
+                
+                needToShuffleRegs = true;
+            }
+        }
+
+        if (needToShuffleRegs)
```

Comment:
I had to write all this code because the equivalent code in CodegenXarcxh.cpp only handled the case where we had exactly two registers and it did so using simplistic enumerate all possibilities implementation.